### PR TITLE
Product taxon position don't affect parent taxon - API

### PR DESCRIPTION
**Pull request by everwhatever**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.11 |
| Bug fix?        | yes  |
| New feature?    | no |
| BC breaks?      | no|
| Deprecations?   | no |
| Related tickets |[ticket](https://github.com/Sylius/Sylius/issues/14308) |
| License         | MIT  |

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+use Sylius\Component\Taxonomy\Repository\TaxonRepositoryInterface;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class ProductsByTaxonExtension implements ContextAwareQueryCollectionExtensionInterface
+{
+    public function __construct(private UserContextInterface $userContext,
+                                private TaxonRepositoryInterface $taxonRepository,
+    )
+    {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]:
```suggestion
    public function __construct(
        private UserContextInterface $userContext,
        private TaxonRepositoryInterface $taxonRepository,
    ) {
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use Doctrine\ORM\QueryBuilder;
+use Sylius\Bundle\ApiBundle\Context\UserContextInterface;
+use Sylius\Bundle\ApiBundle\Serializer\ContextKeys;
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Taxonomy\Repository\TaxonRepositoryInterface;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class ProductsByTaxonExtension implements ContextAwareQueryCollectionExtensionInterface
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php:23]: You didn't add a spec for this class

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        if (!is_a($resourceClass, ProductInterface::class, true)) {
+            return;
+        }
+
+        $user = $this->userContext->getUser();
+        if ($user !== null && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+            return;
+        }
+
+        if (isset($context['filters'])){
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]:
```suggestion
        if (isset($context['filters'])) {
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            $taxonCode = $context['filters']['productTaxons.taxon.code'] ?? null;
+            $this->addSortingToQuery($queryBuilder, $taxonCode);
+        }
+    }
+
+    private function addSortingToQuery(QueryBuilder $queryBuilder, ?string $taxonCode): void
+    {
+        if (!$taxonCode){
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]:
```suggestion
        if (!$taxonCode) {
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            return;
+        }
+
+        $taxon = $this->taxonRepository->findOneBy(['code' => $taxonCode]);
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]: Change the query below to not use the taxon repository in this extension 🤔

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+        if (!$taxon) {
+            return;
+        }
+
+        $queryBuilder
+            ->leftJoin('o.productTaxons', 'productTaxons', 'WITH', 'productTaxons.taxon = :taxonId')
+            ->orderBy('productTaxons.position', 'ASC')
+            ->setParameter('taxonId', $taxon->getId())
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]: Use `$queryNameGenerator` to generate the parameter name

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        $taxon = $this->taxonRepository->findOneBy(['code' => $taxonCode]);
+
+        if (!$taxon) {
+            return;
+        }
+
+        $queryBuilder
+            ->leftJoin('o.productTaxons', 'productTaxons', 'WITH', 'productTaxons.taxon = :taxonId')
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]: Use a root alias `$rootAlias = $queryBuilder->getRootAliases()[0];` instead of `o`

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 9]
```diff
@@ -47,29 +45,25 @@ public function applyToCollection(
...
-        if (!$taxon) {
-            return;
-        }
+        $rootAlias = $queryBuilder->getRootAliases()[0];
 
         $queryBuilder
-            ->leftJoin('o.productTaxons', 'productTaxons', 'WITH', 'productTaxons.taxon = :taxonId')
-            ->orderBy('productTaxons.position', 'ASC')
-            ->setParameter('taxonId', $taxon->getId())
+            ->leftJoin(sprintf('%s.productTaxons', $rootAlias), 'pt', 'WITH', sprintf('pt.product = %s.id', $rootAlias))
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]: `pt` and `tax` should also be properly aliased (`$queryNameGenerator->generateParameterName`)

File: features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature [hunk 10]
```diff
@@ -18,14 +18,14 @@ Feature: Viewing products from specific taxon sorted by position
         And this product is in "Toys" taxon at 2nd position
         And this product is in "Soft Toys" taxon at 2nd position
 
-    @ui
+    @api @ui
     Scenario: Seeing sorted product on list from a specific taxon
-        When I browse products from taxon "Toys"
+        When I browse products from product taxon code "Toys"
```
Comment (GSadee) [inline features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature:23]: Why did you change the step definition?

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+                'WITH',
+                sprintf('%s.product = %s.id', $productTaxonAliasName, $rootAlias))
+            ->leftJoin(sprintf('%s.taxon', $productTaxonAliasName),
+                $taxonAliasName,
+                'WITH',
+                sprintf('%s.code = :%s', $taxonAliasName, $taxonCodeParameterName))
+            ->orderBy(sprintf('%s.position', $productTaxonAliasName),
+                'ASC')
+            ->setParameter($taxonCodeParameterName, $taxonCode)
+        ;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]:
```suggestion
        $queryBuilder
            ->leftJoin(
                sprintf('%s.productTaxons', $rootAlias),
                $productTaxonAliasName,
                'WITH',
                sprintf('%s.product = %s.id', $productTaxonAliasName, $rootAlias))
            ->leftJoin(
                sprintf('%s.taxon', $productTaxonAliasName),
                $taxonAliasName,
                'WITH',
                sprintf('%s.code = :%s', $taxonAliasName, $taxonCodeParameterName))
            ->orderBy(sprintf('%s.position', $productTaxonAliasName), 'ASC')
            ->setParameter($taxonCodeParameterName, $taxonCode)
        ;
```
or
```suggestion
        $queryBuilder
            ->leftJoin(sprintf('%s.productTaxons', $rootAlias), $productTaxonAliasName)
            ->leftJoin(sprintf('%s.taxon', $productTaxonAliasName), $taxonAliasName)
            ->andWhere(sprintf('%s.product = %s.id', $productTaxonAliasName, $rootAlias))
            ->andWhere(sprintf('%s.code = :%s', $taxonAliasName, $taxonCodeParameterName))
            ->orderBy(sprintf('%s.position', $productTaxonAliasName), 'ASC')
            ->setParameter($taxonCodeParameterName, $taxonCode)
        ;
```

File: features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature (hunk 10 above)
Comment (everwhatever) [reply features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature:23]: Because previous definition was from this method: `\Sylius\Behat\Context\Api\Shop\ProductContext::iBrowseProductsFromTaxon`
And I needed to change filter 'taxon' to 'productTaxons.taxon.code'. I've created new method because I didn't want to break other tests using `iBrowseProductsFromTaxon` method

Comment (GSadee) [reply features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature:23]: Now I wonder why we have two similar filters defined: `taxon` and `productTaxons.taxon.code`, probably we should have defined only one of them (`taxon` I guess). What is more, there is a filter definition https://github.com/Sylius/Sylius/blob/1.13/src/Sylius/Bundle/ApiBundle/Filter/Doctrine/TaxonFilter.php and you should apply the fix also there and of course test it properly.

Comment (GSadee) [reply features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature:23]: @lchrusciel do you have any opinion on these filters? 👀

Comment (everwhatever) [reply features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature:23]: I've decided to use filter with `productTaxons.taxon.code` because I found it here: https://master.demo.sylius.com/api/v2/docs
In section 'product' there are two params 'taxon' or 'productTaxons.taxon.code' and when I was using 'taxon' it didn't return any product (It worked only with 'productTaxons.taxon.code') so that's why I used second param (filter)

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 12]
```diff
@@ -92,6 +92,19 @@ public function iBrowseProductsFromTaxon(?TaxonInterface $taxon = null): void
         }
     }
 
+    /**
+     * @When I browse products from product taxon code :taxon
+     */
+    public function iBrowseProductsFromProductTaxonCode(?TaxonInterface $taxon = null): void
```
Comment (coldic3) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php]:
```suggestion
    public function iBrowseProductsFromProductTaxonCode(TaxonInterface $taxon): void
```
Is there any case when it can be null? :thinking:

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+        $user = $this->userContext->getUser();
+        if ($user !== null && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+            return;
+        }
+
+        if (isset($context['filters'])) {
+            $taxonCode = $context['filters']['productTaxons.taxon.code'] ?? null;
+            $this->addSortingToQuery($queryBuilder, $taxonCode, $queryNameGenerator);
+        }
```
Comment (coldic3) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]: I think you can make `$taxonCode` not nullable in `addSortingToQuery` method and move the condition with early return to the `applyToCollection`:
```suggestion
        $taxonCode = $context['filters']['productTaxons.taxon.code'] ?? null;
        if (null === $taxonCode) {
            return;
        }
        
        $this->addSortingToQuery($queryBuilder, $taxonCode, $queryNameGenerator);
```

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/ProductsByTaxonExtensionSpec.php [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    function it_does_nothing_if_current_resource_is_not_a_product(
+        UserContextInterface $userContext,
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
+    ): void {
+        $userContext->getUser()->shouldNotBeCalled();
+        $queryBuilder->getRootAliases()->shouldNotBeCalled();
+        $queryBuilder->addSelect('translation')->shouldNotBeCalled();
```
Comment (coldic3) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/ProductsByTaxonExtensionSpec.php]: How does it work? I mean in the `ProductsByTaxonExtension` I can't see that `addSelect` method is called.

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/ProductsByTaxonExtensionSpec.php [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        UserContextInterface $userContext,
+        UserInterface $user,
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
+    ): void {
+        $userContext->getUser()->willReturn($user);
+        $user->getRoles()->willReturn(['ROLE_API_ACCESS']);
+
+        $queryBuilder->getRootAliases()->shouldNotBeCalled();
+        $queryBuilder->addSelect('translation')->shouldNotBeCalled();
```
Comment (coldic3) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/ProductsByTaxonExtensionSpec.php]: Similar as above, this spec seems to be incorrect.

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/ProductsByTaxonExtensionSpec.php [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        UserInterface $user,
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
+    ): void {
+        $userContext->getUser()->willReturn($user);
+        $user->getRoles()->willReturn(['ROLE_API_ACCESS']);
+
+        $queryBuilder->getRootAliases()->shouldNotBeCalled();
+
+        $this->applyToCollection($queryBuilder, $queryNameGenerator, ProductInterface::class, 'get', []);
```
Comment (coldic3) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/ProductsByTaxonExtensionSpec.php]: I think we should use:
```suggestion
        $this->applyToCollection($queryBuilder, $queryNameGenerator, ProductInterface::class, 'get', ['filters' => ['productTaxons.taxon.code' => 't_shirts']]);
```
to be sure we test the correct case.

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+        $taxonCode = $context['filters']['productTaxons.taxon.code'] ?? null;
+        if (null === $taxonCode) {
+            return;
+        }
+
+        $this->addSortingToQuery($queryBuilder, $taxonCode, $queryNameGenerator);
+    }
+
+    private function addSortingToQuery(QueryBuilder $queryBuilder, ?string $taxonCode, QueryNameGeneratorInterface $queryNameGenerator): void
```
Comment (coldic3) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]:
```suggestion
    private function addSortingToQuery(QueryBuilder $queryBuilder, string $taxonCode, QueryNameGeneratorInterface $queryNameGenerator): void
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+        $taxonCodeParameterName = $queryNameGenerator->generateParameterName('taxonCode');
+        $productTaxonAliasName = $queryNameGenerator->generateJoinAlias('productTaxons');
+        $taxonAliasName = $queryNameGenerator->generateJoinAlias('taxon');
+
+        $queryBuilder
+            ->leftJoin(
+                sprintf('%s.productTaxons', $rootAlias),
+                $productTaxonAliasName,
+                'WITH',
+                sprintf('%s.product = %s.id', $productTaxonAliasName, $rootAlias))
```
Comment (coldic3) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]:
```suggestion
                sprintf('%s.product = %s.id', $productTaxonAliasName, $rootAlias)
            )
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+            ->leftJoin(
+                sprintf('%s.taxon', $productTaxonAliasName),
+                $taxonAliasName,
+                'WITH',
+                sprintf('%s.code = :%s', $taxonAliasName, $taxonCodeParameterName))
```
Comment (coldic3) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByTaxonExtension.php]:
```suggestion
                sprintf('%s.code = :%s', $taxonAliasName, $taxonCodeParameterName)
            )
```

File: src/Sylius/Bundle/ApiBundle/spec/Filter/Doctrine/TaxonFilterSpec.php [hunk 20]
```diff
@@ -34,17 +34,25 @@ function it_adds_taxon_filter_if_property_is_taxon(
         QueryBuilder $queryBuilder,
         QueryNameGeneratorInterface $queryNameGenerator,
     ): void {
+        $taxonLeft = 3;
+        $taxonRight = 5;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Filter/Doctrine/TaxonFilterSpec.php]:
```suggestion
```
and use a specific values in methods below

**Review by GSadee — APPROVED** [review]

File: .github/workflows/application.yml [hunk 21]
```diff
@@ -582,6 +582,14 @@ jobs:
...
                     path: etc/build/
                     if-no-files-found: ignore
+                    
+            -   
+                name: Upload varlog logs
+                uses: actions/upload-artifact@v3
+                with:
+                    name: "varlog logs (non-JS, PHP ${{ matrix.php }}, Symfony ${{ matrix.symfony }})"
+                    path: var/log/
+                    if-no-files-found: ignore
```
Comment (GSadee) [inline .github/workflows/application.yml]: I would be for reverting this change as we are not uploading logs for other builds, and I believe it is not needed

File: features/product/viewing_products/viewing_products_from_specific_taxon_sorted_by_position.feature [hunk 22]
```diff
@@ -18,13 +18,19 @@ Feature: Viewing products from specific taxon sorted by position
...
-    Scenario: Seeing sorted product on list from a specific taxon
-        When I browse products from taxon "Toys"
+    @api @ui
+    Scenario: Seeing sorted product on list from a specific taxon code
+        When I browse products from product taxon code "Toys"
         Then I should see 3 products in the list
         And they should have order like "Old pug", "Young pug" and "Small pug"
 
     @ui
+    Scenario: Seeing sorted product on list from a child taxon code
```
Comment (GSadee) [inline features/product/viewing_products/viewing_products_from_specific_taxon_sorted_by_position.feature:28]: Why does this scenario not have an `@api` tag?

**Review by GSadee — COMMENTED** [review]
And squash the last `test` commits or change its descriptions

File: features/product/viewing_products/viewing_products_from_specific_taxon_sorted_by_position.feature [hunk 23]
```diff
@@ -18,13 +18,19 @@ Feature: Viewing products from specific taxon sorted by position
...
-    Scenario: Seeing sorted product on list from a specific taxon
-        When I browse products from taxon "Toys"
+    @api @ui
+    Scenario: Seeing sorted product on list from a specific taxon code
+        When I browse products from product taxon code "Toys"
         Then I should see 3 products in the list
         And they should have order like "Old pug", "Young pug" and "Small pug"
 
-    @ui
+    @ui @api
```
Comment (GSadee) [inline features/product/viewing_products/viewing_products_from_specific_taxon_sorted_by_position.feature:27]: You are not consistent in the order of tags in this PR

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, @everwhatever! :1st_place_medal:
